diff --git a/source/pdf/pdf-fontfile.c b/source/pdf/pdf-fontfile.c
index f0787bc..102d037 100644
--- a/source/pdf/pdf-fontfile.c
+++ b/source/pdf/pdf-fontfile.c
@@ -32,68 +32,131 @@
 #endif
 #endif
 
+// EBD: Add support for dynamically loaded external fonts >>>
+extern char ext_font_Courier[1024];
+extern char ext_font_CourierBold[1024];
+extern char ext_font_CourierOblique[1024];
+extern char ext_font_CourierBoldOblique[1024];
+extern char ext_font_Helvetica[1024];
+extern char ext_font_HelveticaBold[1024];
+extern char ext_font_HelveticaOblique[1024];
+extern char ext_font_HelveticaBoldOblique[1024];
+extern char ext_font_TimesRoman[1024];
+extern char ext_font_TimesBold[1024];
+extern char ext_font_TimesItalic[1024];
+extern char ext_font_TimesBoldItalic[1024];
+extern char ext_font_Symbol[1024];
+extern char ext_font_ZapfDingbats[1024];
+
+extern unsigned char *
+get_bytes_from_file(char *filename, unsigned int *len);
+
 unsigned char *
 pdf_lookup_builtin_font(const char *name, unsigned int *len)
 {
+	unsigned char * font_hexdump = NULL;
+	*len = 0;
 	if (!strcmp("Courier", name)) {
-		*len = sizeof pdf_font_NimbusMon_Reg;
-		return (unsigned char*) pdf_font_NimbusMon_Reg;
+		font_hexdump = get_bytes_from_file(ext_font_Courier, len);
+		if(font_hexdump == NULL) {
+			*len = sizeof pdf_font_NimbusMon_Reg;
+			return (unsigned char*) pdf_font_NimbusMon_Reg;
+		}
 	}
 	if (!strcmp("Courier-Bold", name)) {
-		*len = sizeof pdf_font_NimbusMon_Bol;
-		return (unsigned char*) pdf_font_NimbusMon_Bol;
+		font_hexdump = get_bytes_from_file(ext_font_CourierBold, len);
+		if(font_hexdump == NULL) {
+			*len = sizeof pdf_font_NimbusMon_Bol;
+			return (unsigned char*) pdf_font_NimbusMon_Bol;
+		}
 	}
 	if (!strcmp("Courier-Oblique", name)) {
-		*len = sizeof pdf_font_NimbusMon_Obl;
-		return (unsigned char*) pdf_font_NimbusMon_Obl;
+		font_hexdump = get_bytes_from_file(ext_font_CourierOblique, len);
+		if(font_hexdump == NULL) {
+			*len = sizeof pdf_font_NimbusMon_Obl;
+			return (unsigned char*) pdf_font_NimbusMon_Obl;
+		}
 	}
 	if (!strcmp("Courier-BoldOblique", name)) {
-		*len = sizeof pdf_font_NimbusMon_BolObl;
-		return (unsigned char*) pdf_font_NimbusMon_BolObl;
+		font_hexdump = get_bytes_from_file(ext_font_CourierBoldOblique, len);
+		if(font_hexdump == NULL) {
+			*len = sizeof pdf_font_NimbusMon_BolObl;
+			return (unsigned char*) pdf_font_NimbusMon_BolObl;
+		}
 	}
 	if (!strcmp("Helvetica", name)) {
-		*len = sizeof pdf_font_NimbusSan_Reg;
-		return (unsigned char*) pdf_font_NimbusSan_Reg;
+		font_hexdump = get_bytes_from_file(ext_font_Helvetica, len);
+		if(font_hexdump == NULL) {
+			*len = sizeof pdf_font_NimbusSan_Reg;
+			return (unsigned char*) pdf_font_NimbusSan_Reg;
+		}
 	}
 	if (!strcmp("Helvetica-Bold", name)) {
-		*len = sizeof pdf_font_NimbusSan_Bol;
-		return (unsigned char*) pdf_font_NimbusSan_Bol;
+		font_hexdump = get_bytes_from_file(ext_font_HelveticaBold, len);
+		if(font_hexdump == NULL) {
+			*len = sizeof pdf_font_NimbusSan_Bol;
+			return (unsigned char*) pdf_font_NimbusSan_Bol;
+		}
 	}
 	if (!strcmp("Helvetica-Oblique", name)) {
-		*len = sizeof pdf_font_NimbusSan_Ita;
-		return (unsigned char*) pdf_font_NimbusSan_Ita;
+		font_hexdump = get_bytes_from_file(ext_font_HelveticaOblique, len);
+		if(font_hexdump == NULL) {
+			*len = sizeof pdf_font_NimbusSan_Ita;
+			return (unsigned char*) pdf_font_NimbusSan_Ita;
+		}
 	}
 	if (!strcmp("Helvetica-BoldOblique", name)) {
-		*len = sizeof pdf_font_NimbusSan_BolIta;
-		return (unsigned char*) pdf_font_NimbusSan_BolIta;
+		font_hexdump = get_bytes_from_file(ext_font_HelveticaBoldOblique, len);
+		if(font_hexdump == NULL) {
+			*len = sizeof pdf_font_NimbusSan_BolIta;
+			return (unsigned char*) pdf_font_NimbusSan_BolIta;
+		}
 	}
 	if (!strcmp("Times-Roman", name)) {
-		*len = sizeof pdf_font_NimbusRom_Reg;
-		return (unsigned char*) pdf_font_NimbusRom_Reg;
+		font_hexdump = get_bytes_from_file(ext_font_TimesRoman, len);
+		if(font_hexdump == NULL) {
+			*len = sizeof pdf_font_NimbusRom_Reg;
+			return (unsigned char*) pdf_font_NimbusRom_Reg;
+		}
 	}
 	if (!strcmp("Times-Bold", name)) {
-		*len = sizeof pdf_font_NimbusRom_Med;
-		return (unsigned char*) pdf_font_NimbusRom_Med;
+		font_hexdump = get_bytes_from_file(ext_font_TimesBold, len);
+		if(font_hexdump == NULL) {
+			*len = sizeof pdf_font_NimbusRom_Med;
+			return (unsigned char*) pdf_font_NimbusRom_Med;
+		}
 	}
 	if (!strcmp("Times-Italic", name)) {
-		*len = sizeof pdf_font_NimbusRom_Ita;
-		return (unsigned char*) pdf_font_NimbusRom_Ita;
+		font_hexdump = get_bytes_from_file(ext_font_TimesItalic, len);
+		if(font_hexdump == NULL) {
+			*len = sizeof pdf_font_NimbusRom_Ita;
+			return (unsigned char*) pdf_font_NimbusRom_Ita;
+		}
 	}
 	if (!strcmp("Times-BoldItalic", name)) {
-		*len = sizeof pdf_font_NimbusRom_MedIta;
-		return (unsigned char*) pdf_font_NimbusRom_MedIta;
+		font_hexdump = get_bytes_from_file(ext_font_TimesBoldItalic, len);
+		if(font_hexdump == NULL) {
+			*len = sizeof pdf_font_NimbusRom_MedIta;
+			return (unsigned char*) pdf_font_NimbusRom_MedIta;
+		}
 	}
 	if (!strcmp("Symbol", name)) {
-		*len = sizeof pdf_font_StandardSymL;
-		return (unsigned char*) pdf_font_StandardSymL;
+		font_hexdump = get_bytes_from_file(ext_font_Symbol, len);
+		if(font_hexdump == NULL) {
+			*len = sizeof pdf_font_StandardSymL;
+			return (unsigned char*) pdf_font_StandardSymL;
+		}
 	}
 	if (!strcmp("ZapfDingbats", name)) {
-		*len = sizeof pdf_font_Dingbats;
-		return (unsigned char*) pdf_font_Dingbats;
+		font_hexdump = get_bytes_from_file(ext_font_ZapfDingbats, len);
+		if(font_hexdump == NULL) {
+			*len = sizeof pdf_font_Dingbats;
+			return (unsigned char*) pdf_font_Dingbats;
+		}
 	}
-	*len = 0;
-	return NULL;
+	return font_hexdump;
 }
+// EBD: Add support for dynamically loaded external fonts <<<
 
 unsigned char *
 pdf_lookup_substitute_font(int mono, int serif, int bold, int italic, unsigned int *len)
