diff --git a/source/pdf/pdf-fontfile.c b/source/pdf/pdf-fontfile.c
index 86b6682..67e9af8 100644
--- a/source/pdf/pdf-fontfile.c
+++ b/source/pdf/pdf-fontfile.c
@@ -25,68 +25,132 @@
 #endif
 #endif
 
+// EBD: Add support for dynamically loaded external fonts >>>
+extern char ext_font_Courier[1024];
+extern char ext_font_CourierBold[1024];
+extern char ext_font_CourierOblique[1024];
+extern char ext_font_CourierBoldOblique[1024];
+extern char ext_font_Helvetica[1024];
+extern char ext_font_HelveticaBold[1024];
+extern char ext_font_HelveticaOblique[1024];
+extern char ext_font_HelveticaBoldOblique[1024];
+extern char ext_font_TimesRoman[1024];
+extern char ext_font_TimesBold[1024];
+extern char ext_font_TimesItalic[1024];
+extern char ext_font_TimesBoldItalic[1024];
+extern char ext_font_Symbol[1024];
+extern char ext_font_ZapfDingbats[1024];
+
+extern unsigned char *
+get_bytes_from_file(char *filename, unsigned int *len);
+
+
 unsigned char *
 pdf_lookup_builtin_font(fz_context *ctx, const char *name, unsigned int *len)
 {
+	unsigned char * font_hexdump = NULL;
+	*len = 0;
 	if (!strcmp("Courier", name)) {
-		*len = sizeof pdf_font_NimbusMono_Regular;
-		return (unsigned char*) pdf_font_NimbusMono_Regular;
+		font_hexdump = get_bytes_from_file(ext_font_Courier, len);
+		if(font_hexdump == NULL) {
+			*len = sizeof pdf_font_NimbusMono_Regular;
+			return (unsigned char*) pdf_font_NimbusMono_Regular;
+		}
 	}
 	if (!strcmp("Courier-Bold", name)) {
-		*len = sizeof pdf_font_NimbusMono_Bold;
-		return (unsigned char*) pdf_font_NimbusMono_Bold;
+		font_hexdump = get_bytes_from_file(ext_font_CourierBold, len);
+		if(font_hexdump == NULL) {
+			*len = sizeof pdf_font_NimbusMono_Bold;
+			return (unsigned char*) pdf_font_NimbusMono_Bold;
+		}
 	}
 	if (!strcmp("Courier-Oblique", name)) {
-		*len = sizeof pdf_font_NimbusMono_Oblique;
-		return (unsigned char*) pdf_font_NimbusMono_Oblique;
+		font_hexdump = get_bytes_from_file(ext_font_CourierOblique, len);
+		if(font_hexdump == NULL) {
+			*len = sizeof pdf_font_NimbusMono_Oblique;
+			return (unsigned char*) pdf_font_NimbusMono_Oblique;
+		}
 	}
 	if (!strcmp("Courier-BoldOblique", name)) {
-		*len = sizeof pdf_font_NimbusMono_BoldOblique;
-		return (unsigned char*) pdf_font_NimbusMono_BoldOblique;
+		font_hexdump = get_bytes_from_file(ext_font_CourierBoldOblique, len);
+		if(font_hexdump == NULL) {
+			*len = sizeof pdf_font_NimbusMono_BoldOblique;
+			return (unsigned char*) pdf_font_NimbusMono_BoldOblique;
+		}
 	}
 	if (!strcmp("Helvetica", name)) {
-		*len = sizeof pdf_font_NimbusSanL_Reg;
-		return (unsigned char*) pdf_font_NimbusSanL_Reg;
+		font_hexdump = get_bytes_from_file(ext_font_Helvetica, len);
+		if(font_hexdump == NULL) {
+			*len = sizeof pdf_font_NimbusSanL_Reg;
+			return (unsigned char*) pdf_font_NimbusSanL_Reg;
+		}
 	}
 	if (!strcmp("Helvetica-Bold", name)) {
-		*len = sizeof pdf_font_NimbusSanL_Bol;
-		return (unsigned char*) pdf_font_NimbusSanL_Bol;
+		font_hexdump = get_bytes_from_file(ext_font_HelveticaBold, len);
+		if(font_hexdump == NULL) {
+			*len = sizeof pdf_font_NimbusSanL_Bol;
+			return (unsigned char*) pdf_font_NimbusSanL_Bol;
+		}
 	}
 	if (!strcmp("Helvetica-Oblique", name)) {
-		*len = sizeof pdf_font_NimbusSanL_RegIta;
-		return (unsigned char*) pdf_font_NimbusSanL_RegIta;
+		font_hexdump = get_bytes_from_file(ext_font_HelveticaOblique, len);
+		if(font_hexdump == NULL) {
+			*len = sizeof pdf_font_NimbusSanL_RegIta;
+			return (unsigned char*) pdf_font_NimbusSanL_RegIta;
+		}
 	}
 	if (!strcmp("Helvetica-BoldOblique", name)) {
-		*len = sizeof pdf_font_NimbusSanL_BolIta;
-		return (unsigned char*) pdf_font_NimbusSanL_BolIta;
+		font_hexdump = get_bytes_from_file(ext_font_HelveticaBoldOblique, len);
+		if(font_hexdump == NULL) {
+			*len = sizeof pdf_font_NimbusSanL_BolIta;
+			return (unsigned char*) pdf_font_NimbusSanL_BolIta;
+		}
 	}
 	if (!strcmp("Times-Roman", name)) {
-		*len = sizeof pdf_font_NimbusRomNo9L_Reg;
-		return (unsigned char*) pdf_font_NimbusRomNo9L_Reg;
+		font_hexdump = get_bytes_from_file(ext_font_TimesRoman, len);
+		if(font_hexdump == NULL) {
+			*len = sizeof pdf_font_NimbusRomNo9L_Reg;
+			return (unsigned char*) pdf_font_NimbusRomNo9L_Reg;
+		}
 	}
 	if (!strcmp("Times-Bold", name)) {
-		*len = sizeof pdf_font_NimbusRomNo9L_Med;
-		return (unsigned char*) pdf_font_NimbusRomNo9L_Med;
+		font_hexdump = get_bytes_from_file(ext_font_TimesBold, len);
+		if(font_hexdump == NULL) {
+			*len = sizeof pdf_font_NimbusRomNo9L_Med;
+			return (unsigned char*) pdf_font_NimbusRomNo9L_Med;
+		}
 	}
 	if (!strcmp("Times-Italic", name)) {
-		*len = sizeof pdf_font_NimbusRomNo9L_RegIta;
-		return (unsigned char*) pdf_font_NimbusRomNo9L_RegIta;
+		font_hexdump = get_bytes_from_file(ext_font_TimesItalic, len);
+		if(font_hexdump == NULL) {
+			*len = sizeof pdf_font_NimbusRomNo9L_RegIta;
+			return (unsigned char*) pdf_font_NimbusRomNo9L_RegIta;
+		}
 	}
 	if (!strcmp("Times-BoldItalic", name)) {
-		*len = sizeof pdf_font_NimbusRomNo9L_MedIta;
-		return (unsigned char*) pdf_font_NimbusRomNo9L_MedIta;
+		font_hexdump = get_bytes_from_file(ext_font_TimesBoldItalic, len);
+		if(font_hexdump == NULL) {
+			*len = sizeof pdf_font_NimbusRomNo9L_MedIta;
+			return (unsigned char*) pdf_font_NimbusRomNo9L_MedIta;
+		}
 	}
 	if (!strcmp("Symbol", name)) {
-		*len = sizeof pdf_font_StandardSymL;
-		return (unsigned char*) pdf_font_StandardSymL;
+		font_hexdump = get_bytes_from_file(ext_font_Symbol, len);
+		if(font_hexdump == NULL) {
+			*len = sizeof pdf_font_StandardSymL;
+			return (unsigned char*) pdf_font_StandardSymL;
+		}
 	}
 	if (!strcmp("ZapfDingbats", name)) {
-		*len = sizeof pdf_font_Dingbats;
-		return (unsigned char*) pdf_font_Dingbats;
+		font_hexdump = get_bytes_from_file(ext_font_ZapfDingbats, len);
+		if(font_hexdump == NULL) {
+			*len = sizeof pdf_font_Dingbats;
+			return (unsigned char*) pdf_font_Dingbats;
+		}
 	}
-	*len = 0;
-	return NULL;
+	return font_hexdump;
 }
+// EBD: Add support for dynamically loaded external fonts <<<
 
 unsigned char *
 pdf_lookup_substitute_font(fz_context *ctx, int mono, int serif, int bold, int italic, unsigned int *len)
