diff --git a/include/mupdf/fitz/context.h b/include/mupdf/fitz/context.h
index 35f4f74f..6bd4661d 100644
--- a/include/mupdf/fitz/context.h
+++ b/include/mupdf/fitz/context.h
@@ -125,6 +125,9 @@ struct fz_context_s
 	fz_tuning_context *tuning;
 	fz_document_handler_context *handler;
 	fz_output_context *output;
+	// EBD: context flags >>>
+	int ebookdroid_nightmode;
+	// EBD: context flags <<<
 };
 
 /*
diff --git a/source/fitz/draw-device.c b/source/fitz/draw-device.c
index 25323b15..4aa04295 100644
--- a/source/fitz/draw-device.c
+++ b/source/fitz/draw-device.c
@@ -1344,6 +1344,16 @@ fz_draw_fill_image(fz_context *ctx, fz_device *devp, fz_image *image, const fz_m
 			}
 		}
 
+		// EBD: inverted mode >>>
+		if (ctx->ebookdroid_nightmode) {
+			fz_pixmap *inverted = fz_new_pixmap(ctx, pixmap->colorspace, pixmap->w, pixmap->h, pixmap->alpha);
+			memcpy(inverted->samples, pixmap->samples, pixmap->stride * pixmap->h);
+			fz_invert_pixmap(ctx, inverted);
+			fz_drop_pixmap(ctx, pixmap);
+			pixmap = inverted;
+		}
+		// EBD: inverted mode <<<
+
 		fz_paint_image(state->dest, &state->scissor, state->shape, pixmap, &local_ctm, alpha * 255, !(devp->hints & FZ_DONT_INTERPOLATE_IMAGES), devp->flags & FZ_DEVFLAG_GRIDFIT_AS_TILED);
 
 		if (state->blendmode & FZ_BLEND_KNOCKOUT)
